### DISCUSSION TITLE
Building nedit-ng on Mac Catalina

### DISCUSSION THREAD

**Discussion by grege2** [eteran/nedit-ng#170] [General]
Apologies if there is a nedit-ng mailing list I should be using here.

I've finally upgraded to Catalina and good old 32-bit Nedit on X11 & Motif & XQuartz has stopped working. I've been recommended to nedit-ng, by Markku Ilmanen, thanks Markku.

I'm building away. Got cmake ok. I have a bison on the Mac but looks a bit old, 2.3, I'm happy to upgrade that.

I installed Macports, and it can install boost but the pre-requisites are a bit fearsome:

$ sudo port install boost
--->  Computing dependencies for boost
The following dependencies will be installed: 
 bzip2
 expat
 gettext
 icu
 libedit
 libffi
 libiconv
 lz4
 lzma
 ncurses
 openssl
 python38
 python3_select
 python_select
 sqlite3
 xz
 zlib
 zstd
Continue? [Y/n]: 

I'm just checking if anyone has any comment on installing all that list. Some are already on my Mac, like bzip2, I don't know why port wants to install that. Some look like big installs, eg. Python 3.8. I have a Python 3.3 on my system, but it's years old.

The Mac has Qt but seems to be rev 4.4, not sure if that will do.

Appreciate any comments ! As a quid pro quo, I will document in the FAQ how to get nedit-ng going on Mac Catalina. 

Thanks, Greg E.

**Comment by anjohnson** [comment]
Hi Greg,

I have successfully used Homebrew to install the required dependencies, and its list of dependencies for Boost (1.74.0) is only icu4c. It may be that the Macports version of Boost has everything enabled, but you shouldn't need that for Nedit-ng, is there a lesser Boost package available? Homebrew has Qt-5.15.1, so I would suggest switching to that.

- Andrew

**Comment by grege2** [comment]
Hi Andrew, many thanks, I will try your suggestions.
Greg.

On Tue, 3 Nov 2020 at 11:25, Andrew Johnson <notifications@github.com>
wrote:

> Hi Greg,
>
> I have successfully used Homebrew to install the required dependencies,
> and its list of dependencies for Boost (1.74.0) is only icu4c. It may be
> that the Macports version of Boost has everything enabled, but you
> shouldn't need that for Nedit-ng, is there a lesser Boost package
> available? Homebrew has Qt-5.15.1, so I would suggest switching to that.
>
>    - Andrew
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/eteran/nedit-ng/issues/170#issuecomment-720802406>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AH66LHCP3XFD4W36567ASPLSN5EY3ANCNFSM4TICQCVA>
> .
>


-- 
*Greg Edwards*

**Comment by eteran** [comment]
I am also considering upping the nedit-ng requirements to use c++17, that should make ditching boost at least a **possibility** and hopefully make non-linux installs slightly easier

**Comment by grege2** [comment]
Woo hoo! I have got it working on the Mac (Catalina 10.15.7). Imported my old nedit.rc pretty well, my macros seem ok at first glance. Colour scheme imported ok. Fonts need some work but that is well signalled in the build and doco. I only need one font anyway.

Homebrew was the key to a simple install of Bison, Boost, Qt - thanks @anjohnson   The build was error free, a few "deprecated" warnings.

I have to do other stuff for a while now, but thought I'd just report in. I'll document all this in Wiki soon, after I give it some more exercising. 

Cheers.

**Comment by anjohnson** [comment]
Congrat's @grege2 

You do have to change Apple's `⌘H` key-binding to be able to use "Find Selection" in Nedit-ng, otherwise your window disappears every time you try to do a search! In System Preferences go to Keyboard, Shortcuts, App Shortcuts and add an All Applications shortcut for `Hide nedit-ng`. I set it to Shift+Option+Command+H `⌥⇧⌘H` but I never normally use these "Hide <app>" keystrokes anyway so almost anything obscure would work for me.

The main thing I'd like to change on the Mac version would be to switch the key-bindings to use the normal Mac bindings for things like word-left/right, using the ⌥/alt/option key instead of ⌘/command – switching between Nedit-ng and other Mac windows causes momentary confusion and I find I'm constantly using the wrong modifier keys. There are a few other minor bugs which are probably differences in what Qt does and they don't affect me very often.

**Comment by eteran** [comment]
@grege2 fantastic. Please do document it in the wiki, and of course, let me know if the wiki doesn't have the right permissions.

**Comment by grege2** [comment]
Hi, quick progress update, but I don't want to send too much trivia. Thanks @anjohnson for the keyboard hints, I got ⌘H re-mapped for nedit-ng, so that the old Ctrl-H "Find Selection" works (as of course does Shift ⌘H, "backward find selection"). Actually I never use ⌘H to hide a window, I just minimise it with the yellow dot, so that was a good find. ( Mac doesn't seem to have a window manager binding for "send window to the back", or "send window back one layer", I used to find that useful on old X window managers like mwm. )

I don't use "forward_word" or "backward_word" much so I wouldn't have noticed the issue there. 

I've got my appearance how I want it .. get ready for a shock .. I've had this style for about 30 years .. it started on Emacs and then Jot on Irix at SGI ..

![scheme](https://user-images.githubusercontent.com/33416604/98307917-041ac080-201b-11eb-9703-32d6f8c18710.png)

I could use a 15 or 16 point Courier font, nedit-ng jumps from 14 to 18. My screen is 1680 x 1050, Macbook Pro 15", so the sweet spot for me is about 15 or 16 point font.

nedit-ng is very sweet indeed. Trivia: love the X "close tab" on every tab, I've been wanting that for years, but it clashes totally with X/Motif. Love dragging the tabs around too.

Ok I'm going to use it seriously on my main project at the moment, a 12,000 line piece of JavaScript ( a p5.js app, https://p5js.org/ ), my little hobby. I'll keep a log of observations and get back to you soon. 

Cheers, GE.

**Comment by eteran** [comment]
Looks fantastic. I think my code asks the system for the font sizes, but I can double-check that code to see if an in-between size should be offered.

You can though... manually edit the config file to set the size to any number you like to try it out though  ;-)

**Comment by grege2** [comment]
Hi all. Thanks @eteran for the font tip, I have tried "16" or "17" as font size in config.ini and it looks good.

I also installed Nedit-ng on my wife's fairly virgin Macbook on Mojave 10.14.6 and all went well. I've documented the Mojave and Catalina installs in a new Wiki page. Please review and change anything you like.

Cheers, GE

**Comment by eteran** [comment]
Fantastic, I'll take a look at the wiki tomorrow. I don't own a mac, so as long as they seem sensible, I'll have to assume that they are correct 🙂

**Comment by anjohnson** [comment]
I didn't have to install bison, GNU Bison 2.3 is installed at `/usr/bin/bison` on both my macOS machines and that's what I used with no problems. Other than that potential simplification the Wiki page looks good to me.

Next job would be to [create a formula and package](https://docs.brew.sh/Adding-Software-to-Homebrew) Nedit-ng for Homebrew. Not something that I have done before but I think it should be relatively straightforward.

**Reply by anjohnson** [reply]
Note that my comment about Bison is no longer true, nedit-ng versions by 2022 need a newer Bison than the one that Apple includes.

**Reply by eteran** [reply]
It's worth noting that a pre built parser is in the source tree so bison isn't strictly needed.

We could just make macOS builds use that.

**Reply by anjohnson** [reply]
I don't think it's worth changing the build for that, Homebrew has Bison and it's only one more package to install. I just added the above note in case someone came across my original comment and hadn't read the Wiki page.

**Comment by grege2** [comment]
Thanks Andrew - I didn't try the build with the built-in Bison 2.3, probably should have. I know nothing about Bison, but a bit of Googling reveals it's a sucessor to Yacc, and there's a newer thing called ANTLR. Bison, Yacc (Yak), Antlr, Gnu - people have been having fun there.

**Comment by grege2** [comment]
Hi Evan @eteran. Going along nicely with Nedit-ng on Mac. I'm having a few minor gremlins with key bindings, and some of my old macros. But they could be my own fault, so I've stripped down my (20 years old) nedit.rc to the basics, and gradually reintroducing stuff. 

What I'd like to do is import the latest macro or key-binding I'm working on with nedit -import, but not Save it, until it's working ok, so I can keep a gradually growing core of good macros, and debug the newer ones. Does that sound ok ? I never actually used old X/Motif nedit's "-import" option.

Does nedit -import work ok, ie. as expected, the (properly formatted) macros and key bindings are added to / merged with the existing database ? I see from some older issues here that there was a bug or two earlier on, but now resolved. 

Is this issue a good place to keep discussing these Mac config / macro / key-binding things or should I open a new Issue ? Cheers, Greg E.

**Comment by eteran** [comment]
Import should work the way you expect.

Basically it is a way to overlay new settings on top of existing settings merging them together.

**Comment by eteran** [comment]
I would say that if you encounter specific issues, please do open up other tickets so that they can be tracked independently.

Thanks!

**Comment by grege2** [comment]
I think this issue is basically solved, my build of nedit-ng for Mac is working, and I've documented it in the Wiki. My nedit.rc is largely working, ditto my autoload.nm. Now I'm getting some more old key-bindings going but that's not a fundamental build issue. Thanks for all the help ! Closing this one.

**Comment by eteran** [comment]
@grege2 @anjohnson 

I'd like to introduce you to github discussions!

**Reply by anjohnson** [reply]
I haven't used Discussions before, but it does seem to make a bit more sense than a long-running Issue conversation.

**Comment by eteran** [comment]
It's a new feature that is most suitable for more back and forth, especially things which have an answer that cna be flagged as "this in the answer".

**Comment by grege2** [comment]
Thanks guys, I'm on hliday for a couple of weeks, I guess I'll rebuild
latest Nedit-ng on latest Big Sur when I'm back and check things.
(Sadly my Macbbok Pro is 8 years old now, no Monterey for me).

On Wed, 16 Mar 2022 at 09:57, Andrew Johnson ***@***.***>
wrote:

> I don't think it's worth changing the build for that, Homebrew has Bison
> and it's only one more package to install. I just added the above note in
> case someone came across my original comment and hadn't read the Wiki page.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/eteran/nedit-ng/discussions/170#discussioncomment-2368635>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AH66LHCA3TU2YQZZFWSOY5TVAEIUNANCNFSM4U4R3IXA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
*Greg Edwards*

**Comment by anjohnson** [comment]
Something changed in my build today of 2025.2_rc1-15 on MacOS Sequoia 15.5, which might have been deliberate or else something changed upstream although I'm not sure when as it's been a while since I last updated.

The 3 binaries are now each built into their own `*.app` directory structure in the Build tree, and the `make install` target which cmake generated now copies the whole `.app` area into `$(PREFIX)` (not very useful if PREFIX is set to somewhere like `/usr/local`). It's possible to drag those app directories into the Finder's Applications area and the OS recognizes them, but asking it to open a file with them only launches the program without the file.

I have modified my Makefile to copy the binaries itself from inside those `.app` directories so the result is just as it used to be. I'll try to update the Wiki page when I get a moment.
